tests/service_opts_test.go

```diff
@@ -0,0 +1,194 @@
+	updatedYml := getTechnicalEmailsYaml(testProject, name, testPrimaryCloudName, false)
+
+	// Applies updated manifest
+	require.NoError(t, s.Apply(updatedYml))
+
+	// Waits kube objects
+	require.NoError(t, s.GetRunning(grafana, name))
+
+	// THEN
+	// Technical emails are removed from service
+	grafanaAvnUpdated, err := avnClient.Services.Get(ctx, testProject, name)
+	require.NoError(t, err)
+	assert.Empty(t, grafanaAvnUpdated.TechnicalEmails)
+}
+
+func getConnInfoSecretTargetDisabledYaml(project, name, cloudName string, connInfoSecretTargetDisabled *bool) string {
+	baseYaml := `
+apiVersion: aiven.io/v1alpha1
+kind: Grafana
+metadata:
+  name: %[2]s
+spec:
+  authSecretRef:
+    name: aiven-token
+    key: token
+
+  project: %[1]s
+  cloudName: %[3]s
+  plan: startup-1
+`
+	if connInfoSecretTargetDisabled != nil {
+		baseYaml += fmt.Sprintf(`
+  connInfoSecretTargetDisabled: %t
+`, *connInfoSecretTargetDisabled)
+	}
+
+	return fmt.Sprintf(baseYaml, project, name, cloudName)
+}
+
+func runTest(t *testing.T, scenario TestScenario) {
+	defer recoverPanic(t)
+
+	// GIVEN
+	name := randName("grafana")
+	yml := getConnInfoSecretTargetDisabledYaml(testProject, name, testPrimaryCloudName, scenario.initialConnInfoSecretTargetDisabled)
+	s := NewSession(k8sClient, avnClient, testProject)
+
+	// Cleans test afterwards
+	defer s.Destroy()
+
+	// WHEN
+	// Applies given manifest
+	require.NoError(t, s.Apply(yml))
+
+	// Waits kube objects
+	grafana := new(v1alpha1.Grafana)
+	require.NoError(t, s.GetRunning(grafana, name))
+
+	// THEN
+	// Check if service connection info secret is created
+	secret, err := s.GetSecret(grafana.GetName())
+	if scenario.shouldSecretBeCreated {
+		assert.NoError(t, err)
+		assert.NotEmpty(t, secret)
+	} else {
+		assert.Nil(t, secret)
+	}
+
+	// WHEN
+	// Trying to update service an existing service with updated `connInfoSecretTargetDisabled`
+	updatedYml := getConnInfoSecretTargetDisabledYaml(testProject, name, testPrimaryCloudName, scenario.updatedConnInfoSecretTargetDisabled)
+
+	// Applies updated manifest
+	err = s.Apply(updatedYml)
+	expectedErrorMsg := fmt.Sprintf(scenario.expectedErrorMsg, name)
+	assert.EqualErrorf(t, err, expectedErrorMsg, "Error should be: %v, got: %v", expectedErrorMsg, err)
+}
+
+type TestScenario struct {
+	name                                string
+	initialConnInfoSecretTargetDisabled *bool
+	updatedConnInfoSecretTargetDisabled *bool
+	shouldSecretBeCreated               bool
+	expectedErrorMsg                    string
+}
+
+func TestServiceConnInfoSecretTargetDisabled(t *testing.T) {
+	cases := []TestScenario{
+		{
+			// Modifying `connInfoSecretTargetDisabled` from missing to true --> A secret is created, but the update fails
+			// because `connInfoSecretTargetDisabled` can only be set during service creation.
+			name:                                "modify from missing to true",
+			initialConnInfoSecretTargetDisabled: nil,
+			updatedConnInfoSecretTargetDisabled: pointer.Bool(true),
```

Comment:
nit: Probably generating yaml in-place might make it easier to read and debug (less if/else, pointers)
```
baseYaml := getConnInfoSecretTargetDisabledYaml(project, name, cloudName)
disabledTrue := fmt.Sprintf("%s\n  connInfoSecretTargetDisabled: true", baseYaml)
disabledFalse := fmt.Sprintf("%s\n  connInfoSecretTargetDisabled: false", baseYaml)

...
{
name: "no disabled",
yaml: baseYaml,
},

{
name: "with disabled true",
yaml: disabledTrue,
},
{
name: "with disabled false",
yaml: disabledFalse,
},
```

Reply:
This becomes a bit difficult because the name will conflict. I agree that generating yaml in place is better. I have updated the PR with another variant, let me know if this is better.